test/common/stats/allocator_impl_test.cc

```diff
@@ -75,6 +77,63 @@ TEST_F(AllocatorImplTest, GaugesWithSameName) {
   EXPECT_EQ(0, g2->value());
 }
 
+// Test for a race-condition where we may decrement the ref-count of a stat to
+// zero at the same time as we are allocating another instance of that
+// stat. This test reproduces that race organically by having a 12 threads each
+// iterate 10k times.
+TEST_F(AllocatorImplTest, RefCountDecAllocRaceOrganic) {
+  StatName counter_name = makeStat("counter.name");
+  StatName gauge_name = makeStat("gauge.name");
+  Thread::ThreadFactory& thread_factory = Thread::threadFactoryForTest();
+
+  const uint32_t num_threads = 12;
+  const uint32_t iters = 10000;
+  std::vector<Thread::ThreadPtr> threads;
+  absl::Notification go;
+  for (uint32_t i = 0; i < num_threads; ++i) {
+    threads.push_back(thread_factory.createThread([&]() {
+      go.WaitForNotification();
+      for (uint32_t i = 0; i < iters; ++i) {
+        alloc_.makeCounter(counter_name, "", std::vector<Tag>());
+        alloc_.makeGauge(gauge_name, "", std::vector<Tag>(), Gauge::ImportMode::NeverImport);
+      }
+    }));
+  }
+  go.Notify();
+  for (uint32_t i = 0; i < num_threads; ++i) {
+    threads[i]->join();
+  }
+}
+
+// Tests the same scenario as RefCountDecAllocRaceOrganic, but using just two
+// threads and the ThreadSynchronizer, in one iteration. Note that if the code
+// has the bug in it, this test fails fast as expected. However, if the bug is
+// fixed, the allocator's mutex will cause the second thread to block in
+// makeCounter() until the first thread finishes destructing the object. Thus
+// the test gives thread2 5 seconds to complete before releasing thread 1 to
+// complete its destruction of the counter.
+TEST_F(AllocatorImplTest, RefCountDecAllocRaceSynchronized) {
+  StatName counter_name = makeStat("counter.name");
+  Thread::ThreadFactory& thread_factory = Thread::threadFactoryForTest();
+  alloc_.sync().enable();
+  alloc_.sync().waitOn(AllocatorImpl::DecrementToZeroSyncPoint);
+  absl::Notification alloc_done;
+  Thread::ThreadPtr thread1 = thread_factory.createThread([&]() {
+    CounterSharedPtr counter1 = alloc_.makeCounter(counter_name, "", std::vector<Tag>());
+    alloc_done.Notify();
+    counter1->inc();
+    counter1->reset(); // Blocks in thread synchronizer waiting on DecrementToZeroSyncPoint
+  });
+
+  alloc_done.WaitForNotification();
```

Comment:
I don't think you need the Notification now, right? You will just wait on the barrier?

Reply:
good call; done.